gpu/gpu_info_cuda.c

```diff
@@ -82,121 +86,283 @@ void cuda_init(char *cuda_lib_path, cuda_init_resp_t *resp) {
-  }
 
-  ret = (*h.nvmlDeviceGetCount_v2)(&resp->count);
+  ret = (*h.nvmlDeviceGetCount_v2)(&numDevices);
   if (ret != NVML_SUCCESS) {
     snprintf(buf, buflen, "unable to get device count: %d", ret);
-    resp->err = strdup(buf);
-    return;
+    *err = strdup(buf);
+    return 0;
   }
 
-  resp->total = 0;
-  resp->free = 0;
-  for (i = 0; i < resp->count; i++) {
+  deviceMap = new_device_map(numDevices);
+
+  for(int i = 0;i < numDevices;i++) {
+    unsigned int currentMode = 0,pendingMode;
+
     ret = (*h.nvmlDeviceGetHandleByIndex)(i, &device);
     if (ret != NVML_SUCCESS) {
       snprintf(buf, buflen, "unable to get device handle %d: %d", i, ret);
-      resp->err = strdup(buf);
-      return;
+      *err = strdup(buf);
+      free_device_map(deviceMap);
+      return 0;
     }
 
-    ret = (*h.nvmlDeviceGetMemoryInfo)(device, &memInfo);
-    if (ret != NVML_SUCCESS) {
-      snprintf(buf, buflen, "device memory info lookup failure %d: %d", i, ret);
-      resp->err = strdup(buf);
-      return;
+    ret = (*h.nvmlDeviceGetMigMode)(device,&currentMode,&pendingMode);
+    if (ret != NVML_SUCCESS && ret != NVML_ERROR_NOT_SUPPORTED) {
+      snprintf(buf, buflen, "unable to get MIG mode for device %d: %d", i, ret);
+      *err = strdup(buf);
+      free_device_map(deviceMap);
+      return 0;
     }
-    if (h.verbose) {
-      nvmlBrandType_t brand = 0;
-      // When in verbose mode, report more information about
-      // the card we discover, but don't fail on error
-      ret = (*h.nvmlDeviceGetName)(device, buf, buflen);
-      if (ret != RSMI_STATUS_SUCCESS) {
-        LOG(h.verbose, "nvmlDeviceGetName failed: %d\n", ret);
-      } else {
-        LOG(h.verbose, "[%d] CUDA device name: %s\n", i, buf);
-      }
-      ret = (*h.nvmlDeviceGetBoardPartNumber)(device, buf, buflen);
-      if (ret != RSMI_STATUS_SUCCESS) {
-        LOG(h.verbose, "nvmlDeviceGetBoardPartNumber failed: %d\n", ret);
-      } else {
-        LOG(h.verbose, "[%d] CUDA part number: %s\n", i, buf);
-      }
-      ret = (*h.nvmlDeviceGetSerial)(device, buf, buflen);
-      if (ret != RSMI_STATUS_SUCCESS) {
-        LOG(h.verbose, "nvmlDeviceGetSerial failed: %d\n", ret);
-      } else {
-        LOG(h.verbose, "[%d] CUDA S/N: %s\n", i, buf);
-      }
-      ret = (*h.nvmlDeviceGetVbiosVersion)(device, buf, buflen);
-      if (ret != RSMI_STATUS_SUCCESS) {
-        LOG(h.verbose, "nvmlDeviceGetVbiosVersion failed: %d\n", ret);
-      } else {
-        LOG(h.verbose, "[%d] CUDA vbios version: %s\n", i, buf);
-      }
-      ret = (*h.nvmlDeviceGetBrand)(device, &brand);
-      if (ret != RSMI_STATUS_SUCCESS) {
-        LOG(h.verbose, "nvmlDeviceGetBrand failed: %d\n", ret);
-      } else {
-        LOG(h.verbose, "[%d] CUDA brand: %d\n", i, brand);
+    else if(h.verbose && ret != NVML_ERROR_NOT_SUPPORTED) LOG(h.verbose,"MIG Mode is %d\n", currentMode);
+
+    deviceMap->layout[i][0] = device;
+    deviceMap->layout[i][1] = 0;
+
+    if (currentMode == 0x1) {
```

Comment:
We should probably define our own macros in `gpu_info_cuda.h` for the 2 constants for readability 
```c
#define NVML_DEVICE_MIG_DISABLE 0x0
#define NVML_DEVICE_MIG_ENABLE 0x1
```

Reply:
Yea in the docs those are supposed to exist, but apparently not.  Maybe there's a .h that needs to be included.

Reply:
We try to avoid including headers like this in these cgo files otherwise everyone who tries to build ollama from source would need them on their system, so we've adopted the pattern of defining just the bare minimum symbols in our cgo native files to use the libraries.

Reply:
ok this is an east update by me, will add those things.